.openshift-ci/end.sh

```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
+            # Track cluster destroy failures as overall job failures.
+            outcome="${OUTCOME_FAILED}"
+            ;;
+        "${OUTCOME_PASSED}-${OUTCOME_PASSED}-${OUTCOME_PASSED}")
+            # Perfect score!
+            outcome="${OUTCOME_PASSED}"
+            ;;
+        *)
+            info "ERROR: unexpected state in end.sh: ${combined}"
+            outcome="${OUTCOME_FAILED}"
+            ;;
+    esac
+
+    export OVERALL_JOB_OUTCOME="${outcome}"
+    info "Overall job outcome: ${outcome}"
+}
+
+_JUNIT_CLUSTER_CLASS="Cluster"
+_JUNIT_CLUSTER_CREATE_DESCRIPTION="Create"
+_JUNIT_CLUSTER_DESTROY_DESCRIPTION="Destroy"
+
+# generate_cluster_junit() - generate junit records for cluster create & destroy
+# pass & fail.
+generate_cluster_junit() {
+    # Change the presentation for cluster types
+    local cluster_flavor_variant="${CLUSTER_FLAVOR_VARIANT:-unknown}"
+    if [[ ${#cluster_flavor_variant} -le 5 ]]; then
+        # uppercase GKE, ARO, ROSA, etc
+        cluster_flavor_variant="${cluster_flavor_variant^^}"
+    else
+        cluster_flavor_variant="${cluster_flavor_variant/openshift/OpenShift}"
+    fi
+
+    local cluster_create_debug=""
+    if [[ -f "${SHARED_DIR}/cluster_create_failure_debug.txt" ]]; then
+        cluster_create_debug="$(summarize_cluster_debug "${SHARED_DIR}/cluster_create_failure_debug.txt")"
+    else
+        cluster_create_debug="See build.log and Artifacts for details"
+    fi
+
+    if [[ "${CREATE_CLUSTER_OUTCOME:-}" == "${OUTCOME_PASSED}" ]]; then
+        save_junit_success "${_JUNIT_CLUSTER_CLASS}" "${_JUNIT_CLUSTER_CREATE_DESCRIPTION} ${cluster_flavor_variant}"
+    elif [[ "${CREATE_CLUSTER_OUTCOME:-}" == "${OUTCOME_FAILED}" ]]; then
+        save_junit_failure "${_JUNIT_CLUSTER_CLASS}" "${_JUNIT_CLUSTER_CREATE_DESCRIPTION} ${cluster_flavor_variant}" \
+            "${cluster_create_debug}"
+    fi
+
+    local cluster_destroy_debug=""
+    if [[ -f "${SHARED_DIR}/cluster_destroy_failure_debug.txt" ]]; then
+        cluster_destroy_debug="$(summarize_cluster_debug "${SHARED_DIR}/cluster_destroy_failure_debug.txt")"
+    else
+        cluster_destroy_debug="See build.log and Artifacts for details"
+    fi
+
+    if [[ "${DESTROY_CLUSTER_OUTCOME:-}" == "${OUTCOME_PASSED}" ]]; then
+        save_junit_success "${_JUNIT_CLUSTER_CLASS}" "${_JUNIT_CLUSTER_DESTROY_DESCRIPTION} ${cluster_flavor_variant}"
+    elif [[ "${DESTROY_CLUSTER_OUTCOME:-}" == "${OUTCOME_FAILED}" && "${CREATE_CLUSTER_OUTCOME:-}" == "${OUTCOME_PASSED}" ]]; then
+        # Only record cluster destroy failures when the create succeeded. Otherwise it is mostly noise.
+        save_junit_failure "${_JUNIT_CLUSTER_CLASS}" "${_JUNIT_CLUSTER_DESTROY_DESCRIPTION} ${cluster_flavor_variant}" \
+            "${cluster_destroy_debug}"
+    fi
+}
+
+_NUM_LINE_OF_INTEREST=10
+
+summarize_cluster_debug() {
+    file="$1"
+
+    last_lines="$(tail --lines="${_NUM_LINE_OF_INTEREST}" "${file}")"
+    if [[ -z "${last_lines}" ]]; then
```

Comment:
```suggestion
    if [[ -z ${last_lines} ]]; then
```
`[[` does not need quotes and some linters complain about them

Reply:
Hope [this link](https://www.man7.org/linux/man-pages/man1/bash.1.html#:~:text=The%20words%20between%20the%20%5B%5B%20and%20%5D%5D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20do%20not%20undergo%20word%20splitting%20and%20pathname%20expansion.) will work in your browser. If not, search for the third hit of `[[`.